ethereum/contracts/Bridge.sol

```diff
@@ -0,0 +1,75 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.6.2;
+
+import "./Decoder.sol";
+import "./Application.sol";
+
+contract Bridge {
+    using Decoder for bytes;
+
+    uint64 public mostRecentBlock;
+    mapping(uint64 => uint64) public mostRecentBlockEvent;
+    mapping(address => bool) public applications;
+
+
+    constructor(address[] memory apps) public {
+        mostRecentBlock = 0;
+        for(uint256 i = 0; i < apps.length; i++) {
+            if(verifyApp(apps[i])) {
+                applications[apps[i]] = true;
+            }
+        }
+    }
+
+    /**
+     * @dev routes the message to the specified application ID after verifying the operator's signature
+     * @param _message address _message expected type: Message { AppID [32]byte, Payload []byte }
+     * @param _verification bytes _verification expected type: Verification { BlockNumber int64, EventID int64 }
+     */
+    function handle(bytes memory _message, bytes memory _verification)
+        public
+    {
+        uint64 blockNumber = uint64(_verification.slice(0, 8).decodeUint256());
+        require(blockNumber >= mostRecentBlock, "Blocks must be processed chronologically");
+
+        uint64 eventID = uint64(_verification.slice(8, 16).decodeUint256());
+        require(eventID > mostRecentBlockEvent[blockNumber], "Events must be processed chronologically");
```

Comment:
Do we need to guard against `mostRecentBlockEvent[blockNumber]` raising a NotFound error (or however Solidity signals such errors)?

Reply:
Also we'll need to investigate what an Substrate event id represents. It may not be the same as an Ethereum event id which typically represents the numeric index of an event with a block.

I'll circle back to you on this.

Reply:
The null value of `uint` is 0, so an empty `mostRecentBlockEvent[blockNumber]` returns 0; this line is why I requested that the embedded event ID counter starts from 1 on the substrate side. The first event in a block will be event 1 and therefore pass this check.

Reply:
Sounds good 👍